ratelimits/limiter.go

```diff
@@ -83,6 +87,96 @@ type Decision struct {
 	// theoretical arrival time (TAT) of next request. It must be no more than
 	// (burst * (period / count)) in the future at any single point in time.
 	newTAT time.Time
+
+	// transaction is the Transaction that resulted in this Decision. It is
+	// included for the production of verbose Subscriber-facing errors. It is
+	// set by the Limiter before returning the Decision.
+	transaction Transaction
+}
+
+// RateLimitError returns a verbose Subscriber-facing error for denied
+// *Decisions. If the *Decision is allowed, it returns nil.
+func (d *Decision) RateLimitError(clk clock.Clock) error {
+	if d.Allowed {
+		return nil
+	}
+
+	// Add 0-3% jitter to the RetryIn duration to prevent thundering herd.
+	jitter := time.Duration(float64(d.RetryIn) * 0.03 * rand.Float64())
+	retryAfter := d.RetryIn + jitter
+	retryAfterTs := clk.Now().Add(retryAfter).Format(time.RFC3339)
+
+	switch d.transaction.limit.name {
+	case NewRegistrationsPerIPAddress:
+		return berrors.RegistrationsPerIPError(
+			retryAfter,
+			"too many new registrations (%d) from this IP address in the last %s, retry after %s",
+			d.transaction.limit.Burst,
+			d.transaction.limit.Period.Duration,
+			retryAfterTs,
+		)
+
+	case NewRegistrationsPerIPv6Range:
+		return berrors.RateLimitError(
+			retryAfter,
+			"too many new registrations (%d) from this /48 block of IPv6 addresses in the last %s, retry after %s",
+			d.transaction.limit.Burst,
+			d.transaction.limit.Period.Duration,
+			retryAfterTs,
+		)
+	case NewOrdersPerAccount:
+		return berrors.RateLimitError(
+			retryAfter,
+			"too many new orders (%d) from this account in the last %s, retry after %s",
+			d.transaction.limit.Burst,
+			d.transaction.limit.Period.Duration,
+			retryAfterTs,
+		)
+
+	case FailedAuthorizationsPerDomainPerAccount:
+		// Uses bucket key 'enum:regId:domain'.
+		idx := strings.LastIndex(d.transaction.bucketKey, ":")
+		if idx == -1 {
+			return berrors.InternalServerError("empty bucket key while generating error")
+		}
+		domain := d.transaction.bucketKey[idx+1:]
```

Comment:
An alternative here would be to use [strings.Split](https://pkg.go.dev/strings#Split). I think the result is slightly clearer, and I think it gives more robustness against the format of the bucket key changing, but that's definitely just a matter of opinion:
```suggestion
		parts := strings.Split(d.transaction.bucketKey, ":")
		if len(parts) != 3 {
			return berrors.InternalServerError("unrecognized bucket key while generating error")
		}
		domain := parts[2]
```

Reply:
I think I'm going to keep this the way it is because I want it to be consistent with what we're doing on line 156 where the same solution covers both CertificatesPerDomain, CertificatesPerDomainPerAccount.